driver/pipeline/control.h

```diff
@@ -16,55 +16,123 @@
+/// implements retry handling on connection loss and handles temporary hardware errors.
+/// The pipeline forks a thread to repeatedly read from the streamer and write to the sink.
+class Control {
+public:
     Control() = default;
 
+    /// @brief constructs a new control pipeline that opens streamers on a Synnax database
+    /// cluster.
+    /// @param client the Synnax client to use for opening streamers.
+    /// @param streamer_config the ocnfiguration for the Synnax streamer.
+    /// @param sink the sink to write data to. See the Sink interface for more information.
+    /// @param breaker_config the configuration for the breaker used to manage the
+    /// control thread lifecycle and retry requests on connection loss or temporary
+    /// hardware errors.
     Control(
-        std::shared_ptr<task::Context> ctx,
+        std::shared_ptr<synnax::Synnax> client,
         synnax::StreamerConfig streamer_config,
-        std::unique_ptr<Sink> sink,
+        std::shared_ptr<Sink> sink,
         const breaker::Config &breaker_config
     );
 
-    ~Control(){
-        LOG(INFO) << "Control destructor called";
-    }
+    //// @brief constructs a new control pipeline that opens streamers using the given
+    /// streamer factory.
+    /// @param streamer_factory the streamer factory to use for opening streamers.
+    /// @param streamer_config the configuration for opening the streamer.
+    /// @param sink the sink to write data to. See the Sink interface for more information.
+    /// @param breaker_config the configuration for the breaker used to manage the
+    /// control thread lifecycle and retry requests on connection loss or temporary
+    /// hardware errors.
+    Control(
+        std::shared_ptr<StreamerFactory> streamer_factory,
+        synnax::StreamerConfig streamer_config,
+        std::shared_ptr<Sink> sink,
+        const breaker::Config &breaker_config
+    );
 
-    
+    /// @brief starts the control pipeline if it has not already been started. start is
+    /// idempotent, and is safe to call multiple times without stopping the pipeline.
     void start();
-    void stop();
 
+    /// @brief stops the control pipeline, blocking until the control thread has exited.
+    /// If the pipeline has already stopped, this method will return immediately.
+    void stop();
 private:
-    std::shared_ptr<task::Context> ctx;
-
-    /// @brief writer thread.
     std::unique_ptr<std::thread> thread;
-
-    /// @brief synnax writer
-    std::unique_ptr<synnax::Streamer> streamer;
-    synnax::StreamerConfig streamer_config;
-
-    /// @brief synnax writer
-    synnax::WriterConfig writer_config;
-
-    /// @brief daq interface
-    std::unique_ptr<Sink> sink;
-
-    /// @brief breaker
+    std::shared_ptr<StreamerFactory> factory;
+    synnax::StreamerConfig config;
+    std::shared_ptr<Sink> sink;
     breaker::Breaker breaker;
 
     void runInternal();
 
+    void maybeJoinThread() const;
```

Comment:
I don't like this name. The idea of this function should be that all joinable threads should be joined if this function returns (unless the run thread itself calls it which should not be done but is checked for sanity)